Adding new user.addNotification method with Mongodb Update

Hus274 opened the pull request:
[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Habitica_Git#Pull_Request for more info)

[//]: # (Put Issue # or URL here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #8264 

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)
Added a new method to the user schema that uses mongo's update functionality to push the notification to the database.

Currently the method is called addNotificationUpdate which I don't love. If you have a suggestion let me know. Other things I thought of were 
* addNotificationAndSave
* pushNewNotification
* updateNotificationsWithNew

Let me know what you think.

[//]: # (Put User ID in here - found in Settings -> API)

----
UUID: 2a4d30ee-4833-420e-be7c-ad6a6e261c5f

Hus274 committed:
Adding new method to user schema that pushes a new notification to the database with an update operation instead of a save. fixes #8264

khipkin commented on test/api/v3/unit/models/user.test.js:
```diff
@@ -60,7 +60,22 @@ describe('User Model', () => {
     });
 
-    it('can add notifications without data', () => {
+    it('can add notifications using update', async() => {
```
Suggestion: "without data", just for clarity's sake.

Hus274 replied:
yep, meant to have that, updated.

Hus274 committed:
fixing test text

paglias commented on website/server/models/user/methods.js:
```diff
@@ -42,6 +42,15 @@ schema.methods.addNotification = function addUserNotification (type, data = {})
+    type,
+    data,
+  };
+  this.notifications.push(newNotification);
```
What we want to do with this method is to provide a way to add notifications without having to save the user. So this should probably be a static method of the user model (User.statics.addUserNotificationUpdate) that takes the notification data and one or more ids of users. If you issue the update command and add the data to the user doc (this.notifications.push) the next time the user is saved, the notification will be added again to it.

TheHollidayInn replied:
Yea, you don't need to push this at all. Your update query doesn't use it.

Hus274 replied:
Yeah that was a mistake on my part. I was thinking about keeping the local version in sync with the database version and forgot that the database won't handle the deduping on a later save.

TheHollidayInn commented:
I suggest pushNotification. The function doesn't need to detail anything about the internals like your other options do.

Hus274 committed:
changing the addUserNotificationUpdate method to a static method as requested.

Hus274 commented:
I believe I've made all the changes that were requested here. Let me know if you have any more feedback!

Hus274 committed:
Renaming to push notification

TheHollidayInn commented on website/server/models/user/methods.js:
```diff
@@ -35,13 +35,38 @@ schema.methods.sendMessage = async function sendMessage (userToReceiveMessage, m
 };
 
+/**
+ * Creates a notification and based on the input parameters and adds it to the local user notifications array.
```
Change "Creates a notification and based" to "Creates a notification based"

Hus274 replied:
Good catch!

TheHollidayInn commented on website/server/models/user/methods.js:
```diff
@@ -35,13 +35,38 @@ schema.methods.sendMessage = async function sendMessage (userToReceiveMessage, m
+ * Creates a notification based on the type and data input parameters and saves that new notification
+ * to the database directly using an update statement. The local copy of these users are not updated by
+ * this operation.
+ *
```
Can you add a note for use case on this? For example: "Use this function when you want to add a notification to a user(s), but do not have the document(s) opened.

Hus274 replied:
Yeah that's a good call, I'll update it.

Hus274 committed:
fixing comment documentation based on pr comments.

Hus274 commented:
Updated the documentation according to the comments here

paglias commented on website/server/models/user/methods.js:
```diff
@@ -35,13 +35,39 @@ schema.methods.sendMessage = async function sendMessage (userToReceiveMessage, m
+  let newNotification = new UserNotification({type, data});
+  let promises = [];
+  userIds.forEach(userId => {
+    promises.push(this.update({_id: userId}, {$push: {notifications: newNotification}}).exec());
```
@Hus274 I think the only missing change here is that instead of updating each user separately we should use `{_id: {$in: userIds}}` so only one update command is used.

paglias replied:
and to allow a single id to be passed, so the userIds param can be either an array of ids or a single user id string

paglias commented on website/server/models/user/methods.js:
```diff
@@ -35,13 +35,39 @@ schema.methods.sendMessage = async function sendMessage (userToReceiveMessage, m
+ * @param  data  The data to add to the notification
+ */
+schema.statics.pushNotification = async function pushNotification (userIds, type, data = {}) {
+  let newNotification = new UserNotification({type, data});
```
let's validate this to be sure all values are valid, `newNotification.validateSync()`, it returns `undefined` if the doc is valid or the error if it's not, http://mongoosejs.com/docs/api.html#document_Document-validateSync

Hus274 committed:
Changed the update statement to do a multi update and added a validation step. Added tests for both these cases.

Hus274 commented:
Switched this to use the multi update functionality so that a single userId or a list of user Ids can be used. Also added in mongoose validation before saving. Let me know if anything else should be updated here!

paglias commented:
@Hus274 just wanted to say that we'll merge this soon. We're trying to debug some issues with notifications and would prefer to fix that before merging new code.

Hus274 commented:
@paglias sounds good, that makes perfect sense. Thanks for letting me know!

TheHollidayInn commented:
@paglias I believe we have resolved the Notification issue, correct. Should we merge this PR in now?

paglias commented on website/server/models/user/methods.js:
```diff
@@ -35,13 +35,39 @@ schema.methods.sendMessage = async function sendMessage (userToReceiveMessage, m
+ * @param  type  The type of notification to add to the user. Possible values are defined in the UserNotificaiton Schema
+ * @param  data  The data to add to the notification
+ */
+schema.statics.pushNotification = async function pushNotification (userIds, type, data = {}) {
```
since we haven't yet merged this, would it be possible to change the function to allow an entire query to be passed instead of ids only? It'll make the code much easier to use in situations where we want to target users based on some properties instead of ids. I'll do it tomorrow otherwise

Hus274 replied:
yeah, I should be able to make this change today.

Hus274 committed:
Updating pushNotification method to allow a query to be passed instead of an array of userIds to make it more flexible.

Hus274 committed:
Removing createdAt field from tests as it's no longer used.

Hus274 committed:
Merge remote-tracking branch 'upstream/develop' into 8264

Hus274 committed:
Removing only from test suite

Hus274 commented:
@paglias updated to have the user pass a query instead of an array of user ids.

paglias commented:
Thanks! Notated towards your next tier, and welcome to tier 6 :)

Hus274 commented:
@paglias that's awesome!!! Thank you!!!
